lib/Common/Memory/SmallFinalizableHeapBlock.cpp

```diff
@@ -103,6 +141,146 @@ SmallFinalizableHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsi
+        [&](unsigned char _attributes) { *attributes = _attributes; }))
+    {
+        // Couldn't mark children- bail out and come back later
+        this->SetNeedOOMRescan(markContext->GetRecycler());
+    }
+}
+
+template <class TBlockAttributes>
+template <bool doSpecialMark, typename Fn>
+bool SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes>::UpdateAttributesOfMarkedObjects(MarkContext * markContext, void * objectAddress, size_t objectSize, unsigned char attributes, Fn fn)
+{
+    bool noOOMDuringMark = true;
+
+    if (attributes & TrackBit)
+    {
+        Assert((attributes & LeafBit) == 0);
+        IRecyclerVisitedObject* recyclerVisited = static_cast<IRecyclerVisitedObject*>(objectAddress);
+        noOOMDuringMark = markContext->AddPreciselyTracedObject(recyclerVisited);
+
+        if (noOOMDuringMark)
+        {
+            // Object has been successfully processed, so clear NewTrackBit
+            attributes &= ~NewTrackBit;
+        }
+        else
+        {
+            // Set the NewTrackBit, so that the main thread will redo tracking
+            attributes |= NewTrackBit;
+            noOOMDuringMark = false;
+        }
+        fn(attributes);
+    }
+
+#ifdef RECYCLER_STATS
+    RECYCLER_STATS_INTERLOCKED_INC(markContext->GetRecycler(), markData.markCount);
+    RECYCLER_STATS_INTERLOCKED_ADD(markContext->GetRecycler(), markData.markBytes, objectSize);
+
+    // Count track or finalize if we don't have to process it in thread because of OOM.
+    if ((attributes & (TrackBit | NewTrackBit)) != (TrackBit | NewTrackBit))
+    {
+        // Only count those we have queued, so we don't double count
+        if (attributes & TrackBit)
+        {
+            RECYCLER_STATS_INTERLOCKED_INC(markContext->GetRecycler(), trackCount);
+        }
+        if (attributes & FinalizeBit)
+        {
+            // we counted the finalizable object here,
+            // turn off the new bit so we don't count it again
+            // on Rescan
+            attributes &= ~NewFinalizeBit;
+            fn(attributes);
+            RECYCLER_STATS_INTERLOCKED_INC(markContext->GetRecycler(), finalizeCount);
+        }
+    }
+#endif
+
+    return noOOMDuringMark;
+}
+
+// static
+template <class TBlockAttributes>
+bool
+SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes>::RescanObject(SmallRecyclerVisitedHostHeapBlockT<TBlockAttributes> * block, __in_ecount(localObjectSize) char * objectAddress, uint localObjectSize,
+    uint objectIndex, Recycler * recycler)
+{
+    unsigned char const attributes = block->ObjectInfo(objectIndex);
+
+    if ((attributes & TrackBit) != 0)
+    {
+        Assert((attributes & LeafBit) == 0);
+        Assert(block->GetAddressIndex(objectAddress) != SmallHeapBlockT<TBlockAttributes>::InvalidAddressBit);
+
+        if (!recycler->AddPreciselyTracedMark(reinterpret_cast<IRecyclerVisitedObject*>(objectAddress)))
```

Comment:
Why is this not always called when we rescan a visited object?

Reply:
Objects are only pushed onto the preciseStack of the MarkContext if they need traced.  Since the TrackBit currently indicates whether or not we need traced it should guard AddPreciselyTracedMark.